### PR TITLE
Add Access Key login

### PR DESCRIPTION
**Pull request by ravemir**
# One-line summary
Add Access Key login capabilities, since most of functionality only relies on it anyway

## Description
While studying the app for adoption, we found that it uses a "workaround" to allow SSO login (which is what we use)
This is acceptable, but looks slightly more cumbersome+error prone, so we wanted to add an alternative way.
After inspection, we realized that Access Keys are obtained immediately after login with Cookies, indicating that it might be the only credentials used for the remainder of the app.
While this is not strictly true (the "Launch Live Test" option in the "DeviceDetails" component is an example), we can adjust the remaining cases to be off-limits while using this method.

## Types of Changes
- New feature
- Configuration change

## Tasks
_List of tasks you will do to complete the PR_
  - [x] Change UI and API operations to allow interchangeable Access Key+Cookie login for SSO
  - [ ] Adjust Opening of Sauce Labs Sessions from app to not crash when Access Key is used

## Review
_List of tasks the reviewer must do to review the PR_
- [ ] Tests
- [ ] Documentation
- [ ] CHANGELOG

### CODE REVIEW

**Comment by ravemir** [comment]
Once again: please bear with my lack of experience as a Node/React.js developer.

Also: please comment on the usage of Access Keys for this purpose. I have no context on how they are generated, and hence can't make a full assessment on how risky it is for people to copy-paste them around to login every time they need to use this.

File: app/login/components/Login.jsx [hunk 1]
```diff
@@ -237,18 +237,18 @@ export default class Login extends Component<Props> {
...
                     <div className={Styles['login_reset']}>
-                      <span>This app can't use SSO, but a workaround is to get a cookie.</span>{' '}
+                      <span>This app can't use SSO, but a workaround is to get an Access Key or a Cookie.</span>{' '}
                       <span>Check the following {' '}
                         <a
                           className={Styles.link}
                           onClick={() =>
                             shell.openExternal(
-                              'https://github.com/saucelabs/saucelabs-vusb-app/blob/main/docs/SSO.md'
+                              'https://github.com/ravemir/saucelabs-vusb-app/blob/main/docs/SSO.md'
```
Comment (wswebcreation) [inline app/login/components/Login.jsx]: Can you revert this change?

**Review by wswebcreation — COMMENTED** [review]
Hi @ravemir 

Thanks for your contribution, I see where you're going to, the only challenge is that the manual testing screen will not work anymore. 
I currently don't have a different way for this, unless we want to break the manual testing experience for SSO users.
But...., AFAIK you could still set a password as a SSO user. What if you would try to do that, then log in with your username and password (the default login for this tool) and then see if it works?

File: app/login/components/Login.jsx (hunk 1 above)
Comment (ravemir) [reply app/login/components/Login.jsx]: Absolutely: I want my users to see my version of the docs when I launch the app, but that should've stayed on my "customizations" branch.

**Comment by ravemir** [comment]
@wswebcreation : thanks for the lightning fast response. How can I change the password in an SSO account if there's technically no current password that I can insert into the form? I'd be willing to try it out if I could.

(beyond this, however, it would not fit our use-case very well: we'd have dozens of people needing to manually set their passwords the first time, then still need to copy it over the credentials onto a shared machine every time they need it.
In contrast, not starting the session from the GUI is a minor inconvenience, if they can start one from Sauce Labs beforehand.)

Since current versions of "virtual-usb-client.jar" only use Access Keys, maybe there's some other way that can be leveraged using these credentials as well? I will investigate and provide an update if I find something.

For the time being, I committed a fix for broken cookie logins, which also disables the "Launch ... Session" link if an "Access Key" login is used (it's definitely not the best experience, but after wrestling with toggling a Notification component on DevicesOverview, from DeviceDetail, I eventually gave up in favor of a more economical solution).

**Comment by ravemir** [comment]
Update: turns out the links returned by the latest virtual-usb-client.jar only work if they are opened in a browser with the cookie already.

**Comment by ravemir** [comment]
~~Sorry: messed up with this last commit. Please hold while I revert.~~ Everything checks out, now

**Review by wswebcreation — COMMENTED** [review]
Hi @ravemir 

I understand the struggles, but I'm not really _happy_ with the solution that you propose. I need to rethink it. I'm currently also busy with our IAM team to see if we can expose the cookie based on username and accesskey, which would make it much easier for this tool.

I was working on a revamp of the UI and flow which can be found in [this branche](https://github.com/saucelabs/saucelabs-vusb-app/tree/feat/dependency-ts-upgrade) which I haven't touched for almost half a year due to time. That new version doesn't have the login screen and automatically opens a browser instead of a new Electron screen. this will prevent you from providing the cookie and will make live much easier.
Can you check out that branch and see if it works for you? There are still things that I need to optimise and fix, but I might be able to release a pre version of it.

Love to hear your thoughts on that

**Comment by ravemir** [comment]
Worth checking out, but I'm afraid it might still not fit our scenario: since the GUI will run on shared machines (where we want to avoid users do an login to the SSO provider), opening a browser window will not result in a successful login :\

But if exposing the cookie via API is an option, it would definitely solve it with a proper experience all around.

Thank you for your time! If you do reconsider and want to integrate part of these changes, feel free to drop a line and we'll restart this discussion! (e.g. if you want to integrate cookie fetching from the API)

**Comment by wswebcreation** [comment]
Hi @ravemir 

I’m picking up the refactoring after I had contact with our engineers. We are not going to expose the cookie through and API so I will change the flow. My idea would be the following 

- username/accessKey will be the main drivers (will create a product tour on initial startup)
- there will be two ways to connect to a device,  one will be for connecting to an existing session on Sauce Labs, no cookie is needed, the second one will be to start a new session (no cookie needed yet, unless you want to open the shared session through the App)

If you still want to open a session through vUSB GUI, I might add a login screen , but not as a blocking request 

The refactoring will take some time because Electron levelled up their security so I need to put a lot of communication from the renderer thread to the main thread which will cause a lot of effort. I will also implement a simpler design, use latest version of Electron and will start using TypeScript

please let me know if you’d like to see some extra stuff so I can add it

Grtz
Wim

**Comment by ravemir** [comment]
Hello, @wswebcreation ,

Sorry for the (big) delay in replying (competing priorities).

From what I've gathered, with these latest developments you are willing to transition to Username+Access Key logins, foregoing the old cookie scheme completely, correct? If so, this lines up perfectly with the initial request 👍 

Regarding sessions through vUSB GUI:
1. Connecting to an existing session (opened through browser) is essential;
2. Starting a new session is secondary to us if we can't open it, which is the case without a cookie;
3. The only way I see our use case could be covered would be to allow to start the session and have a "Copy to Clipboard" button/link with the URL for the live test (so users could bring it out of the remote session and into their browsers, where a cookie could already exist).

Topic 3 still feels a bit hackish (to have the cookie in the local machine, you still need to login to SauceLabs beforehand and not allow the cookie to expire), but doesn't involve fiddling around with cookies.

In the end, if Username+Access Key is implemented as the main driver, however, I could just let users know that they should start their session beforehand and it would work fine.

Thanks for the time in advance. Please let know what you thought of my proposal and what you end up planning.

**Comment by wswebcreation** [comment]
Hi @ravemir

Thanks for your reply and no problem, I understand the priorities. Same here, I haven't had time in the last few weeks to work on it, but...., I am around 75% of the refactoring which is good.
I'm still thinking on how I will add the `startSession`, which will start a clean session (for which you also need the cookie for). Username and accesskey are the main drivers now. 

Some screens and the flow. I also want to add a product tour in it that will help you easily set up the GUI in a few steps, including some new features for every release (still to be build 😅 )

**Home screen:**
image

**Requirements screen:**
image

**Settings Connection:**
image

**Settings Server:**
image

**Settings Server Proxy:**
image

**Settings Device Proxy:**
image

**Comment by ravemir** [comment]
Looks great! Questions that popped up while looking at the screenshots:
1. Is there going to be an option to forget Username+Access Key credentials? (selfishly looking at our scenario again 🙄)
2. There's no login screen anymore? (wouldn't see it as a problem, as long as the it's obvious that anything failing is due to a missing configuration)

**Comment by wswebcreation** [comment]
Thanks

> Is there going to be an option to forget Username+Access Key credentials? (selfishly looking at our scenario again 🙄)
Can you explain what you mean with that? username and accesskey are needed to communicate with Sauce. The accesskey is not a password and is not connected to the cookie/SSO

> There's no login screen anymore? (wouldn't see it as a problem, as long as the it's obvious that anything failing is due to a missing configuration)
I might introduce it, but in a different way that would not affect SSO users because I then want to introduce an option to select where to open the link

**Comment by ravemir** [comment]
> Can you explain what you mean with that? username and accesskey are needed to communicate with Sauce. The accesskey is not a password and is not connected to the cookie/SSO

I understand that the expected use-case of configuring username+access keys involves doing it only once, after which point it will be saved.
For us, since the troubleshooting machines are shared across several users, and we'd like each user to access Sauce Labs authenticated as themselves, we'd prefer if credentials from a previous user are saved, and accidentally used by subsequent users.
As such, the way to achieve this would be to forget previously used credentials when the app is reopened. Would this be possible with these changes? (I understand if you disagree, but thought I'd share our scenario anyway)

> I might introduce it, but in a different way that would not affect SSO users because I then want to introduce an option to select where to open the link

Understood. No objections here at all, since we'll only be using SSO.

**Comment by wswebcreation** [comment]
@ravemir 

I’ve released a beta version, see https://github.com/saucelabs/saucelabs-vusb-app/releases/tag/v2.0.0-beta

Can you check that one to see if it works for you?
